File: src/main/java/com/codingforcookies/betterrecords/common/util/ClasspathInjector.java
package com.codingforcookies.betterrecords.common.util;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Method;
import java.net.URL;
import java.net.URLClassLoader;

public class ClasspathInjector{

	private static final Class<?>[] parameters = new Class[]{URL.class};

	public static void addFile(File f) throws IOException{
		addURL(f.toURI().toURL());
	}

	public static void addURL(URL url) throws IOException{
		URLClassLoader sysloader = (URLClassLoader) ClassLoader.getSystemClassLoader();
		Class<?> sysclass = URLClassLoader.class;
		try{
			Method method = sysclass.getDeclaredMethod("addURL", parameters);
			method.setAccessible(true);
			method.invoke(sysloader, new Object[]{url});
		}catch(Throwable t){
			t.printStackTrace();
			throw new IOException("Error, could not inject file to system classloader");
		}
	}
}

File: src/main/kotlin/com/codingforcookies/betterrecords/util/ClasspathInjector.kt
package com.codingforcookies.betterrecords.util

import java.io.File
import java.io.IOException
import java.net.URL
import java.net.URLClassLoader

object ClasspathInjector {

    private val parameters = arrayOf<Class<*>>(URL::class.java)

    @Throws(IOException::class)
    fun addFile(f: File) {
        addURL(f.toURI().toURL())
    }

    @Throws(IOException::class)
    fun addURL(url: URL) {
        val sysloader = ClassLoader.getSystemClassLoader() as URLClassLoader
        val sysclass = URLClassLoader::class.java

        try {
            val method = sysclass.getDeclaredMethod("addURL", *parameters)
            method.isAccessible = true
            method.invoke(sysloader, url)
        } catch (t: Throwable) {
            t.printStackTrace()
            throw IOException("Error, could not inject file to system classloader")
        }
    }
}
